components/engine_tirocks/src/write_batch.rs

```diff
@@ -0,0 +1,326 @@
+                .as_inner()
+                .write_multi(&opts, &mut self.wbs[..=self.index])
+                .map_err(r2e)
+        } else {
+            self.engine
+                .as_inner()
+                .write(&opts, &mut self.wbs[0])
+                .map_err(r2e)
+        }
+    }
+
+    fn data_size(&self) -> usize {
+        let mut size = 0;
+        for w in &self.wbs[..=self.index] {
+            size += w.as_bytes().len();
+        }
+        size
+    }
+
+    fn count(&self) -> usize {
+        let mut size = 0;
+        for w in &self.wbs[..=self.index] {
+            size += w.count();
+        }
+        size
+    }
+
+    fn is_empty(&self) -> bool {
+        self.wbs[0].as_bytes().is_empty()
+    }
+
+    fn should_write_to_engine(&self) -> bool {
+        if self.engine.multi_batch_write() {
+            self.index >= WRITE_BATCH_MAX_BATCH
+        } else {
+            self.wbs[0].count() > RocksEngine::WRITE_BATCH_MAX_KEYS
+        }
+    }
+
+    fn clear(&mut self) {
+        for i in 0..=self.index {
+            self.wbs[i].clear();
+        }
+        self.save_points.clear();
+        // Avoid making the wbs too big at one time, then the memory will be kept
+        // after reusing
+        if self.index > WRITE_BATCH_MAX_BATCH {
+            self.wbs.shrink_to(WRITE_BATCH_MAX_BATCH);
+        }
+        self.index = 0;
+    }
+
+    fn set_save_point(&mut self) {
+        self.wbs[self.index].set_save_point();
+        self.save_points.push(self.index);
+    }
+
+    fn pop_save_point(&mut self) -> Result<()> {
+        if let Some(x) = self.save_points.pop() {
+            return self.wbs[x].pop_save_point().map_err(r2e);
+        }
+        Err(engine_traits::Error::Engine(
+            engine_traits::Status::with_error(
+                engine_traits::Code::InvalidArgument,
+                "no save point",
+            ),
+        ))
+    }
+
+    fn rollback_to_save_point(&mut self) -> Result<()> {
+        if let Some(x) = self.save_points.pop() {
+            for i in x + 1..=self.index {
+                self.wbs[i].clear();
+            }
+            self.index = x;
+            return self.wbs[x].rollback_to_save_point().map_err(r2e);
+        }
+        Err(engine_traits::Error::Engine(
+            engine_traits::Status::with_error(
+                engine_traits::Code::InvalidArgument,
+                "no save point",
+            ),
+        ))
+    }
+
+    fn merge(&mut self, mut other: Self) -> Result<()> {
+        if !self.engine.multi_batch_write() {
+            let self_wb = &mut self.wbs[0];
+            for wb in &other.wbs[..=other.index] {
+                self_wb.append(wb).map_err(r2e)?;
+            }
+            return Ok(());
+        }
+        let self_wb = &mut self.wbs[self.index];
+        if self_wb.count() < WRITE_BATCH_LIMIT {
+            self_wb.append(&other.wbs[0]).map_err(r2e)?;
+        }
+        // From this point, either of following statements is true:
```

Comment:
What if `self_wb.count() >= WRITE_BATCH_LIMIT && other.index == 0`? other will be lost.